### ISSUE TITLE
Rubocop server crashing when it attempts to restart

### ISSUE THREAD

**Issue by xxx** [rubocop/rubocop#13445]
## Expected behavior

When running a Rubocop server, I expect that it needs to occasionally restart, and does so seamlessly without crashing or needing additional inputs from me.

## Actual behavior

It crashes on attempting to restart. If I kill the server, it automatically restarts on the next run, and _does_ work until any code change. At that point, the crashes begin again. I see this behavior both on the command line, and in my IDE (Jetbrains RubyMine). Once occurring, _all_ Rubocop commands will crash, including switches like `-V` and `-h`.

We've used Rubocop for years with this setup and haven't any issues like this. My teammates are also not seeing the issue. One potential difference is that I am on Linux and they are on MacOS, but it's unclear why that would suddenly cause a problem after almost 5 years. This feels like an issue specific to my system, but nothing seems to work.

```
RuboCop version incompatibility found, RuboCop server restarting...
/home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/socket_reader.rb:36:in `ensure in read!': undefined method `string' for nil:NilClass (NoMethodError)

        Cache.stderr_path.write(stderr.string)
                                      ^^^^^^^
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/socket_reader.rb:37:in `read!'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/core.rb:90:in `read_socket'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/core.rb:65:in `block in process_input'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/cache.rb:150:in `write_pid_file'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/core.rb:64:in `process_input'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/core.rb:53:in `block in detach_server'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/core.rb:46:in `fork'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/core.rb:46:in `detach_server'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/core.rb:38:in `start'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/client_command/start.rb:42:in `block in run'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/cache.rb:137:in `acquire_lock'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/client_command/start.rb:29:in `run'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/cli.rb:93:in `run_command'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/cli.rb:83:in `process_arguments'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/cli.rb:49:in `run'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/exe/rubocop:8:in `<top (required)>'
        from /home/mpd/.asdf/installs/ruby/3.2.2/bin/rubocop:25:in `load'
        from /home/mpd/.asdf/installs/ruby/3.2.2/bin/rubocop:25:in `<top (required)>'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/cli/exec.rb:58:in `load'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/cli/exec.rb:58:in `kernel_load'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/cli/exec.rb:23:in `run'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/cli.rb:484:in `exec'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/cli.rb:31:in `dispatch'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/cli.rb:25:in `start'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/exe/bundle:48:in `block in <top (required)>'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/friendly_errors.rb:103:in `with_friendly_errors'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/exe/bundle:36:in `<top (required)>'
        from /home/mpd/git-sources/double_loop/bin/bundle:113:in `load'
        from /home/mpd/git-sources/double_loop/bin/bundle:113:in `<main>'
/home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/socket_reader.rb:27:in `read!': uninitialized constant RuboCop::Server::SocketReader::StringIO (NameError)

        stderr = StringIO.new
                 ^^^^^^^^
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/core.rb:90:in `read_socket'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/core.rb:65:in `block in process_input'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/cache.rb:150:in `write_pid_file'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/core.rb:64:in `process_input'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/core.rb:53:in `block in detach_server'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/core.rb:46:in `fork'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/core.rb:46:in `detach_server'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/core.rb:38:in `start'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/client_command/start.rb:42:in `block in run'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/cache.rb:137:in `acquire_lock'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/client_command/start.rb:29:in `run'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/cli.rb:93:in `run_command'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/cli.rb:83:in `process_arguments'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/cli.rb:49:in `run'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/exe/rubocop:8:in `<top (required)>'
        from /home/mpd/.asdf/installs/ruby/3.2.2/bin/rubocop:25:in `load'
        from /home/mpd/.asdf/installs/ruby/3.2.2/bin/rubocop:25:in `<top (required)>'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/cli/exec.rb:58:in `load'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/cli/exec.rb:58:in `kernel_load'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/cli/exec.rb:23:in `run'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/cli.rb:484:in `exec'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/cli.rb:31:in `dispatch'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/cli.rb:25:in `start'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/exe/bundle:48:in `block in <top (required)>'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/friendly_errors.rb:103:in `with_friendly_errors'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/exe/bundle:36:in `<top (required)>'
        from /home/mpd/git-sources/double_loop/bin/bundle:113:in `load'
        from /home/mpd/git-sources/double_loop/bin/bundle:113:in `<main>'
bundler: failed to load command: rubocop (/home/mpd/.asdf/installs/ruby/3.2.2/bin/rubocop)
/home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/client_command/exec.rb:53:in `status': RuboCop server: Could not find status file at: /home/mpd/.cache/rubocop_cache/server/home+mpd+git-sources+double_loop/status (RuntimeError)
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/lib/rubocop/server/client_command/exec.rb:27:in `run'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.66.1/exe/rubocop:12:in `<top (required)>'
        from /home/mpd/.asdf/installs/ruby/3.2.2/bin/rubocop:25:in `load'
        from /home/mpd/.asdf/installs/ruby/3.2.2/bin/rubocop:25:in `<top (required)>'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/cli/exec.rb:58:in `load'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/cli/exec.rb:58:in `kernel_load'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/cli/exec.rb:23:in `run'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/cli.rb:484:in `exec'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/cli.rb:31:in `dispatch'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/cli.rb:25:in `start'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/exe/bundle:48:in `block in <top (required)>'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/lib/bundler/friendly_errors.rb:103:in `with_friendly_errors'
        from /home/mpd/.asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/gems/bundler-2.3.7/exe/bundle:36:in `<top (required)>'
        from /home/mpd/.asdf/installs/ruby/3.2.2/bin/bundle:25:in `load'
        from /home/mpd/.asdf/installs/ruby/3.2.2/bin/bundle:25:in `<main>'
```

## Steps to reproduce the problem

1. Run rubocop when server is _not_ already running. (Should start the server, and run normally)
2. Make any code change.
3. Run rubocop again. Crashes forever.

## RuboCop version

1.68.0 (the trace above shows 1.66.1, but I did update everything to verify - it crashes in the exact same way)
I've also tried purging all rubocop-family gems, then bundle installing, but no. This really feels like something on my system,
but I am very confused as to what it could be. Any hints on debugging this?

The `/home/mpd/.cache/rubocop_cache/server/home+mpd+git-sources+double_loop/status` file mentioned _is_ missing, but all the other files are in that directory (i.e. `lock`, `pid`, `port`, `stderr`, `token`, `version`). I don't know where the `status` file comes from, or why it's missing.

Include the output of `rubocop -V` or `bundle exec rubocop -V` if using Bundler.

```
1.66.1 (using Parser 3.3.6.0, rubocop-ast 1.32.3, running on ruby 3.2.2) [x86_64-linux]
  - rubocop-factory_bot 2.25.1
  - rubocop-graphql 1.5.4
  - rubocop-i18n 1.14.6
  - rubocop-performance 1.21.1
  - rubocop-rails 2.25.1
  - rubocop-rspec 2.29.2
  - rubocop-rspec_rails 2.28.3
  - rubocop-thread_safety 0.5.1
```

`rubocop-graphql`, `rubocop-performance`, and `rubocop-rails` have all had patch updates since everything was last working for me, but I tried a full purge-and-rollback, and it still failed, indicating to me that it might be some sysconfig issue, but no idea what it could be all of a sudden.

**Comment by Earlopain** [comment]
I can't reproduce this, can you show the stacktrace from 1.68? There have been two PRs which should have prevented this error: https://github.com/rubocop/rubocop/pull/13234 and https://github.com/rubocop/rubocop/pull/13379

**Comment by xxx** [comment]
Ok, this time I purged, updated to 1.68, and killed the 1.66.1 server by hand. At this point, 1.68 runs and _no longer crashes_ on code changes. I think before I had not killed the 1.66.1 server manually before trying with 1.68.

tl; dr I think this is resolved. Sorry for the falsish alarm.

**Comment by Earlopain** [comment]
Yes, at that point you can't even close it anymore with `rubocop --stop-server` and have to kill it manually. I'm not too deep into the server to explain why that happens but it does :shrug:. You probably bumped the `psych` gem to 5.2.0 which made you run into this.

**Comment by jeffbax** [comment]
> Yes, at that point you can't even close it anymore with `rubocop --stop-server` and have to kill it manually. I'm not too deep into the server to explain why that happens but it does 🤷. You probably bumped the `psych` gem to 5.2.0 which made you run into this.
> 
> Ah, actually, you might have started rubocop 1.66 with the new psych version. That would do it I think.

This is exactly what bit me today, and then went in a very long series of loops trying to find this thread :) Our app uses `standard` so we are stuck waiting for that to update to a newer rubocop, but glad to see a fix is waiting for us eventually.
